File: ImasLiveDB/Models/MasteryScale.swift
import Foundation
import SwiftUI
import UIKit

/// 習熟度の段階の定義 (ラベルと数)。
///
/// **保存されるのは序数だけ** (`user_marks.kind = "mastery"` の `text_value`)。
/// ラベルはここが持ち、設定で差し替えても既存の記録には触らない。段数から導ける規則
/// (次の段 / 重み / 段数を変えたときの寄せ先) と群化・集計は `imas-core` の
/// `domain/mastery.rs` にある。ここは**文言と色だけ**。
struct MasteryScale: Codable, Equatable, Sendable {
    /// 下から順の段ラベル。`labels[0]` が LV.1。
    var labels: [String]

    /// 既定は 3 段。
    ///
    /// 依頼元のツイートは 4 段 (1回聞いた / 耳に馴染んだ / だいたい覚えた / 覚えた) だったが、
    /// 「耳に馴染んだ」が実機で読んだときに言い方として微妙で、さらに**スワイプの
    /// ボタン幅に収まらず途中で切れる** (「耳に馴染」になる)。短く言い切れる 3 段にした。
    /// 語彙の好みは人によるので、設定 (`MasteryScaleSettingsView`) で変えられる。
    ///
    /// **保存される語彙** (`mastery_scale_labels_v1`)。訳さない・変えない。表示は `displayLabels` /
    /// `label(_:)` を通すと、編集していないプリセットは表示言語の訳 (カタログの mastery.preset.*) になる。
    static let defaultLabels = ["聞いた", "覚えた", "完璧"]  // i18n-ignore(storage): 保存する語彙。表示は mastery.preset.steps3.*

    static let standard = MasteryScale(labels: defaultLabels)

    /// 2 段・4 段のプリセットの保存する語彙 (defaultLabels と同じく ja のまま保存する)。
    private static let twoStepLabels = ["聞いた", "覚えた"]  // i18n-ignore(storage): 保存する語彙。表示は mastery.preset.steps2.*
    private static let fourStepLabels = ["聞いた", "だいたい", "覚えた", "完璧"]  // i18n-ignore(storage): 保存する語彙。表示は mastery.preset.steps4.*

    /// 段数を変えるときの出発点。どのラベルも 4 文字以内にして、
    /// スワイプのボタンが切れないようにしてある (訳もカタログの max_len で 4 文字以内)。
    /// チップの名前は `L10n.Mastery.presetName(steps:)`。
    static let presets: [MasteryScale] = [
        MasteryScale(labels: twoStepLabels),
        MasteryScale(labels: defaultLabels),
        MasteryScale(labels: fourStepLabels),
    ]

    /// 段数。core に渡す `steps`。
    var steps: UInt8 { UInt8(clamping: max(1, min(labels.count, 8))) }

    /// 保存値がプリセットの語彙と完全に同じなら、そのプリセットの表示用の文言 (下から順)。
    /// 利用者が 1 つでも書き換えたラベルは nil (そのまま出す)。
    ///
    /// LocalizedStringResource は作った時点の言語で固まるので、static let に置かず毎回作る。
    private static func presetDisplayKeys(for labels: [String]) -> [LocalizedStringResource]? {
        if labels == twoStepLabels {
            return [L10n.Mastery.presetSteps2Level1, L10n.Mastery.presetSteps2Level2]
        }
        if labels == defaultLabels {
            return [L10n.Mastery.presetSteps3Level1, L10n.Mastery.presetSteps3Level2,
                    L10n.Mastery.presetSteps3Level3]
        }
        if labels == fourStepLabels {
            return [L10n.Mastery.presetSteps4Level1, L10n.Mastery.presetSteps4Level2,
                    L10n.Mastery.presetSteps4Level3, L10n.Mastery.presetSteps4Level4]
        }
        return nil
    }

    /// 表示するラベル (下から順)。編集していないプリセットは表示言語の訳、それ以外は保存値のまま。
    /// ja では常に `labels` と同じ。
    var displayLabels: [String] {
        guard let keys = Self.presetDisplayKeys(for: labels) else { return labels }
        return keys.map { String(localized: $0) }
    }

    /// 表示中のラベル (設定画面の入力) から保存する段階を作る。
    /// 訳したプリセットのままなら、そのプリセットの語彙 (ja) で保存する (保存値を言語で変えない)。
    static func storing(displayLabels: [String]) -> MasteryScale {
        presets.first { $0.displayLabels == displayLabels } ?? MasteryScale(labels: displayLabels)
    }

    /// 表示名。`0` は未設定。
    func label(_ level: UInt8) -> String {
        guard level > 0 else { return String(localized: L10n.Mastery.levelUnset) }
        let i = Int(level) - 1
        guard i < labels.count else { return "LV.\(level)" }
        if let keys = Self.presetDisplayKeys(for: labels) { return String(localized: keys[i]) }
        return labels[i]
    }

    /// 一覧のチップに出す短い名前。長いラベルは頭から詰める。
    func shortLabel(_ level: UInt8) -> String {
        let full = label(level)
        return full.count <= 6 ? full : String(full.prefix(5)) + "…"
    }

    /// スワイプで出るボタンのラベル。
    ///
    /// swipe actions のボタンは**中身の幅の合計が画面に収まらないと端が見切れる**。
    /// 「…」を足すとその 1 文字ぶん更に広がるので、ここでは足さずに素で切る。
    /// 4 文字あれば既定の 4 段 (1回聞いた / 耳に馴染んだ / だいたい覚えた / 覚えた) が
    /// iPhone の幅に収まる。
    func swipeLabel(_ level: UInt8) -> String {
        let full = label(level)
        return full.count <= 4 ? full : String(full.prefix(4))
    }
}

// MARK: - 段の色 (単一色相の濃度ランプ)

/// ヒートマップに並べる以上、段ごとに色相を変えない。
/// 色相を変えると「濃い＝進んでいる」がマスの列で読めなくなる。
/// システム accent は塗らない (DS 原則: 色はエンティティ側から来る) ので、
/// 習熟度専用の 1 色を濃度で割る。
enum MasteryPalette {
    /// 未設定は面を持たず、点線の枠だけ。
    static let empty = Color.clear

    private static let rampLight = [0xD7E5F4, 0x9DC4E6, 0x5392CE, 0x1C5FA3]
    private static let rampDark  = [0x17304A, 0x27547F, 0x4287C6, 0x7FBAF0]

    /// `level` (1..=steps) の面の色。段数が 4 未満なら上寄りを使う
    /// (2 段のときに薄い 2 色だけになって「進んだ感」が出ないのを避ける)。
    static func fill(level: UInt8, steps: UInt8) -> Color {
        guard level > 0 else { return empty }
        let idx = rampIndex(level: level, steps: steps)
        return Color(UIColor { tc in
            let hex = tc.userInterfaceStyle == .dark ? rampDark[idx] : rampLight[idx]
            return UIColor(red: CGFloat((hex >> 16) & 0xFF) / 255,
                           green: CGFloat((hex >> 8) & 0xFF) / 255,
                           blue: CGFloat(hex & 0xFF) / 255, alpha: 1)
        })
    }

    /// その面の上に乗せる文字色。薄い段は濃い文字、濃い段は抜き文字。
    static func ink(level: UInt8, steps: UInt8) -> Color {
        guard level > 0 else { return DS.ink3 }
        let idx = rampIndex(level: level, steps: steps)
        return Color(UIColor { tc in
            let dark = tc.userInterfaceStyle == .dark
            // ライトは上位 2 段が濃いので白抜き、ダークは上位が明るいので黒文字。
            if dark { return idx >= 2 ? UIColor(red: 0.02, green: 0.07, blue: 0.12, alpha: 1) : .white }
            return idx >= 2 ? .white : UIColor(red: 0.07, green: 0.16, blue: 0.24, alpha: 1)
        })
    }

    private static func rampIndex(level: UInt8, steps: UInt8) -> Int {
        let s = max(1, Int(steps))
        let l = max(1, min(Int(level), s))
        // 段数が 4 でないときは 4 段のランプ上へ等間隔に写す (最上段は必ず一番濃い)。
        let idx = Int((Double(l - 1) / Double(max(s - 1, 1))) * 3.0 + 0.5)
        return min(max(idx, 0), 3)
    }
}

File: ImasLiveDB-Android/app/src/main/kotlin/com/fugaif/imaslivedb/ui/theme/MasteryScale.kt
package com.fugaif.imaslivedb.ui.theme

import android.content.res.Resources
import androidx.compose.runtime.Composable
import androidx.compose.ui.graphics.Color
import com.fugaif.imaslivedb.i18n.DisplayText
import com.fugaif.imaslivedb.i18n.generated.L10n
import com.fugaif.imaslivedb.i18n.resolve

/**
 * 習熟度の段階の定義 (ラベルと数)。iOS `Models/MasteryScale.swift` の移植。
 *
 * **保存されるのは序数だけ** (`user_marks.kind = "mastery"` の `text_value`)。
 * ラベルはここが持ち、設定で差し替えても既存の記録には触らない。段数から導ける規則
 * (次の段 / 重み / 段数を変えたときの寄せ先) と群化・集計は共有コア
 * (`imas-core` の `domain/mastery.rs`) にある。ここは**文言と色だけ**。
 */
data class MasteryScale(val labels: List<String>) {

    /** 段数。コアに渡す `steps`。 */
    val steps: UByte get() = labels.size.coerceIn(1, 8).toUByte()

    /** 表示名を画面の言語で文字列にしたもの ([labelText] を解決する)。`0` は未設定。 */
    @Composable
    fun label(level: UByte): String = labelText(level).resolve()

    /** 一覧のチップに出す短い名前 (長いラベルは頭から詰める)。 */
    @Composable
    fun shortLabel(level: UByte): String = shorten(label(level))

    /**
     * 表示するラベル (下から順)。編集していないプリセットは画面の言語の訳、それ以外は保存値のまま。
     * ja では常に [labels] と同じ。設定画面の入力欄に出す (iOS `MasteryScale.displayLabels`)。
     */
    fun displayLabels(res: Resources): List<String> =
        presetLabelTexts(labels)?.map { it.resolve(res) } ?: labels

    /**
     * 表示名 (文言の値)。`0` は未設定。
     *
     * 保存しているのは利用者が選んだラベルの文字列そのもの ([AppPreferences.masteryScale])。
     * それが ja のプリセットのまま (書き換えていない) なら今の言語の文言で出し、書き換えていれば
     * 利用者の入力なのでそのまま出す (設計 §9.3。保存の形は変えない)。
     */
    fun labelText(level: UByte): DisplayText {
        if (level.toInt() == 0) return L10n.Mastery.levelUnset
        val i = level.toInt() - 1
        val stored = labels.getOrNull(i) ?: return DisplayText.Verbatim("LV.$level")
        return presetLabelTexts(labels)?.getOrNull(i) ?: DisplayText.Verbatim(stored)
    }

    companion object {
        /**
         * 既定は 3 段。
         *
         * 依頼元は 4 段 (1回聞いた / 耳に馴染んだ / だいたい覚えた / 覚えた) だったが、
         * 「耳に馴染んだ」が読んだときの言い方として微妙で、さらに**行の操作面に
         * 収まらず途中で切れる**。短く言い切れる 3 段にした。語彙の好みは人によるので
         * 設定で変えられる。
         */
        // i18n-ignore(storage): 保存するラベルの既定値 (mastery_scale_labels_v1)。表示は labelText が引く
        val defaultLabels = listOf("聞いた", "覚えた", "完璧")
        val standard = MasteryScale(defaultLabels)

        // i18n-ignore(storage): 選ぶと保存されるラベル (ja のまま保存し、表示は labelText が引く)
        private val twoStepLabels = listOf("聞いた", "覚えた")
        // i18n-ignore(storage): 選ぶと保存されるラベル (ja のまま保存し、表示は labelText が引く)
        private val fourStepLabels = listOf("聞いた", "だいたい", "覚えた", "完璧")

        /**
         * 段数を変えるときの出発点。どれも 4 文字以内にして操作面で切れないようにする
         * (訳もカタログの max_len で 4 文字以内)。チップの名前は [presetNameText]。
         */
        val presets: List<MasteryScale> = listOf(
            MasteryScale(twoStepLabels),
            MasteryScale(defaultLabels),
            MasteryScale(fourStepLabels),
        )

        /** プリセットの名前 (2段 / 3段 / 4段) の文言。steps は段数。 */
        fun presetNameText(steps: Int): DisplayText = L10n.Mastery.presetName(steps = steps)

        /**
         * 表示中のラベル (設定画面の入力) から保存する段階を作る。訳したプリセットのままなら、
         * そのプリセットの語彙 (ja) で保存する (保存値を言語で変えない。iOS `MasteryScale.storing`)。
         */
        fun storing(displayLabels: List<String>, res: Resources): MasteryScale =
            presets.firstOrNull { it.displayLabels(res) == displayLabels } ?: MasteryScale(displayLabels)

        /** 一覧のチップに出す短い名前 (長いラベルは頭から詰める)。 */
        fun shorten(full: String): String = if (full.length <= 6) full else full.take(5) + "…"

        /** 保存しているラベルが ja のプリセットと一致するとき、その段ごとの文言。一致しなければ null。 */
        private fun presetLabelTexts(stored: List<String>): List<DisplayText>? = when (stored) {
            twoStepLabels -> listOf(L10n.Mastery.presetSteps2Level1, L10n.Mastery.presetSteps2Level2)
            defaultLabels -> listOf(
                L10n.Mastery.presetSteps3Level1, L10n.Mastery.presetSteps3Level2,
                L10n.Mastery.presetSteps3Level3,
            )
            fourStepLabels -> listOf(
                L10n.Mastery.presetSteps4Level1, L10n.Mastery.presetSteps4Level2,
                L10n.Mastery.presetSteps4Level3, L10n.Mastery.presetSteps4Level4,
            )
            else -> null
        }
    }
}

/**
 * 段の色。ヒートマップ的に並べる以上、**段ごとに色相を変えない**。
 * 色相を変えると「濃い＝進んでいる」が列で読めなくなる。
 * システムの accent は塗らない (DS 原則: 色はエンティティ側から来る) ので、
 * 習熟度専用の 1 色を濃度で割る。
 */
object MasteryPalette {
    // アプリはダーク基調なので、上の段ほど明るい側へ伸ばす (iOS のダーク用ランプと同値)。
    private val ramp = listOf(
        Color(0xFF17304A), Color(0xFF27547F), Color(0xFF4287C6), Color(0xFF7FBAF0),
    )

    fun fill(level: UByte, steps: UByte): Color =
        if (level.toInt() == 0) Color.Transparent else ramp[rampIndex(level, steps)]

    /** その面の上に乗せる文字色。上位 2 段は明るいので黒文字に倒す。 */
    fun ink(level: UByte, steps: UByte): Color =
        if (level.toInt() == 0) DS.ink3
        else if (rampIndex(level, steps) >= 2) Color(0xFF06121F) else Color.White

    /** 段数が 4 でないときは 4 段のランプ上へ等間隔に写す (最上段は必ず一番濃い)。 */
    private fun rampIndex(level: UByte, steps: UByte): Int {
        val s = steps.toInt().coerceAtLeast(1)
        val l = level.toInt().coerceIn(1, s)
        val idx = ((l - 1).toDouble() / (s - 1).coerceAtLeast(1) * 3.0 + 0.5).toInt()
        return idx.coerceIn(0, 3)
    }
}
